File: 2023/20230117/Rotation.js
function solution(s) {
  let answer = checkBracket(s) ? 1 : 0;
  let result = '';
  for (let i = 1; i < s.length; i += 1) {
    result = s.slice(i) + s.slice(0, i);
    if (checkBracket(result)) answer += 1;
  }
  return answer;
}

function checkBracket(s) {
  const bracket = {
    ')': '(',
    '}': '{',
    ']': '[',
  };
  const stack = [];
  for (let i = 0; i < s.length; i += 1) {
    if (s[i] === '(' || s[i] === '{' || s[i] === '[') stack.push(s[i]);
    else if (stack[stack.length - 1] === bracket[s[i]]) stack.pop();
    else return false;
  }
  return !stack.length;
}

File: 2023/20230116/rotation/src/main/java/Rotation.java
import java.util.Stack;

public class Rotation {
    public int solution(String s) {
        int answer = 0;

        for (int i = 0; i < s.length(); i += 1) {
            String str = s.substring(i, s.length()) + s.substring(0, i);

            answer += validate(str) ? 1 : 0;
        }

        return answer;
    }

    public boolean validate(String str) {
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < str.length(); i += 1) {
            char c = str.charAt(i);

            if (stack.isEmpty()) {
                stack.push(c);
                continue;
            }

            if (c == ')' && stack.peek() == '(') {
                stack.pop();
                continue;
            }

            if (c == '}' && stack.peek() == '{') {
                stack.pop();
                continue;
            }

            if (c == ']' && stack.peek() == '[') {
                stack.pop();
                continue;
            }

            stack.push(c);
        }

        return stack.empty();
    }
}
